// @codesandbox/ab/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.useInitializeExperimentStore = exports.useExperimentResultSync = exports.useExperimentResult = exports.ExperimentValues = exports.initializeExperimentStore = exports.getExperimentResultSync = exports.getExperimentResult = void 0;
var experiment_store_1 = require("./experiment-store");
Object.defineProperty(exports, "getExperimentResult", {
    enumerable: true,
    get: function() {
        return experiment_store_1.getExperimentResult;
    }
});
Object.defineProperty(exports, "getExperimentResultSync", {
    enumerable: true,
    get: function() {
        return experiment_store_1.getExperimentResultSync;
    }
});
Object.defineProperty(exports, "initializeExperimentStore", {
    enumerable: true,
    get: function() {
        return experiment_store_1.initializeExperimentStore;
    }
});
Object.defineProperty(exports, "ExperimentValues", {
    enumerable: true,
    get: function() {
        return experiment_store_1.ExperimentValues;
    }
});
var react_1 = require("./react");
Object.defineProperty(exports, "useExperimentResult", {
    enumerable: true,
    get: function() {
        return react_1.useExperimentResult;
    }
});
Object.defineProperty(exports, "useExperimentResultSync", {
    enumerable: true,
    get: function() {
        return react_1.useExperimentResultSync;
    }
});
Object.defineProperty(exports, "useInitializeExperimentStore", {
    enumerable: true,
    get: function() {
        return react_1.useInitializeExperimentStore;
    }
});
